integration/testnet/network.go

```diff
@@ -533,19 +544,34 @@ func (net *FlowNetwork) AddNode(t *testing.T, bootstrapDir string, nodeConf Cont
 			// uncomment line below to point the access node exclusively to a single collection node
 			// nodeContainer.addFlag("static-collection-ingress-addr", "collection_1:9000")
 			nodeContainer.addFlag("collection-ingress-port", "9000")
-			nodeContainer.opts.HealthCheck = testingdock.HealthCheckCustom(healthcheckAccessGRPC(hostGRPCPort))
-			nodeContainer.Ports[AccessNodeAPIPort] = hostGRPCPort
-			nodeContainer.Ports[AccessNodeAPIProxyPort] = hostHTTPProxyPort
-			net.AccessPorts[AccessNodeAPIPort] = hostGRPCPort
-			net.AccessPorts[AccessNodeAPIProxyPort] = hostHTTPProxyPort
+
+			var accessNodeAPIPortKey, accessNodeAPIProxyPortKey string
+			if nodeConf.Unstaked {
+				accessNodeAPIPortKey = UnstakedAccessNodeAPIPort
+				accessNodeAPIProxyPortKey = UnstakedAccessNodeAPIProxyPort
+			} else {
+				nodeContainer.opts.HealthCheck = testingdock.HealthCheckCustom(healthcheckAccessGRPC(hostGRPCPort))
+				accessNodeAPIPortKey = AccessNodeAPIPort
+				accessNodeAPIProxyPortKey = AccessNodeAPIProxyPort
+			}
+
+			nodeContainer.Ports[accessNodeAPIPortKey] = hostGRPCPort
+			nodeContainer.Ports[accessNodeAPIProxyPortKey] = hostHTTPProxyPort
+			net.AccessPorts[accessNodeAPIPortKey] = hostGRPCPort
+			net.AccessPorts[accessNodeAPIProxyPortKey] = hostHTTPProxyPort
 
 			if nodeConf.ParticipatesInUnstakedNetwork {
 				hostUnstakedPort := testingdock.RandomPort(t)
 				containerUnstakedPort := "9876/tcp"
 				nodeContainer.bindPort(hostUnstakedPort, containerUnstakedPort)
 				nodeContainer.addFlag("unstaked-bind-addr", fmt.Sprintf("%s:9876", nodeContainer.Name()))
-				nodeContainer.Ports[UnstakedNetworkPort] = hostUnstakedPort
-				net.AccessPorts[UnstakedNetworkPort] = hostUnstakedPort
+				if nodeConf.Unstaked {
+					access1 := net.ContainerByName("access_1")
+					access1NodeID := access1.Config.NodeID.String()
+					nodeContainer.addFlag("staked-access-node-id", access1NodeID)
+					nodeContainer.addFlag("staked", "false")
+					nodeContainer.addFlag("staked-access-node-address", fmt.Sprintf("%s:9876", access1.Name()))
+				}
```

Comment:
I think hardcoding the access node ID is fragile and could break in the future.

It might work for this specific test since we create the staked AN first and so it gets the name "access_1" thanks to [this naming logic](https://github.com/onflow/flow-go/blob/066a386e7a23108829a4a66e2ac0fc62a7b6e664/integration/testnet/network.go#L820), but in general I don't think we should rely on the staked AN getting created first.

What if in the future we want to test a setup where multiple staked ANs act as upstreams, and we want each unstaked node to use a specific one?

Instead, let's change the naming logic so that staked AN's are created with name "staked_access_{n}" (similarly for unstaked), and for unstaked nodes we can create a new [node config option](https://github.com/onflow/flow-go/blob/066a386e7a23108829a4a66e2ac0fc62a7b6e664/integration/testnet/network.go#L256-L269) for the upstream node ID which can be populated by a new node options function `WithUpstreamNodeID()`, similar to [this](https://github.com/onflow/flow-go/blob/066a386e7a23108829a4a66e2ac0fc62a7b6e664/integration/testnet/network.go#L302-L306)

Then, the code above can become something like this:

```go
if nodeConf.Unstaked {
	upstream := net.ContainerByID(nodeConf.UpstreamNodeID)
	upstreamNodeID := upstream.Config.NodeID.String()
	nodeContainer.addFlag("staked-access-node-id", upstreamNodeID)
	nodeContainer.addFlag("staked", "false")
	nodeContainer.addFlag("staked-access-node-address", fmt.Sprintf("%s:9876", upstream.Name()))
}
```